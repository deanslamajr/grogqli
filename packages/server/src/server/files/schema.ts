import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
