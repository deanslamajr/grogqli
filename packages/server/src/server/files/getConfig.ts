import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
