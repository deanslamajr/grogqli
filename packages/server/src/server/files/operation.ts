import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import fs from 'fs';

import {
  doesFileExist,
  mapObjectToJsonFile,
  OPERATIONS_NAME_TO_ID_MAPPING_VERSION,
  OPERATIONS_RECORDING_VERSION,
  getOpNameMappingFilePath,
  getOperationRecordingsFilePath,
} from './';

export interface RootTypeRecordingEntry {
  rootTypeId: string;
  recordingId: string;
}

export interface OperationRecording {
  id: string;
  rootTypeRecordings: {
    [rootTypeId: string]: RootTypeRecordingEntry;
  };
}

export type OperationRecordingWithoutId = Omit<OperationRecording, 'id'>;

interface OperationRecordings {
  [opRecordingId: string]: OperationRecording;
}

// pattern to support versioning this file structure
export type OperationRecordingsFile = OperationRecordingsFileVersion1;
export interface OperationRecordingsFileVersion1 {
  id: string;
  version: 1;
  recordings: OperationRecordings;
}

type AddNewRecordingToOperationRecordingsFile = (params: {
  opId: string;
  opRecordingWithoutId: OperationRecordingWithoutId;
}) => Promise<string>;

export const addNewRecordingToOperationRecordingsFile: AddNewRecordingToOperationRecordingsFile = async ({
  opId,
  opRecordingWithoutId,
}) => {
  const pathToOperationRecordingsFile = await getOperationRecordingsFilePath(
    opId
  );
  const opRecordingsFile = editJsonFile(pathToOperationRecordingsFile);

  // if this file has not yet been initialized, throw error
  // this is to prevent unexpected behavior
  if (!doesFileExist(opRecordingsFile)) {
    throw new Error(
      `An operation recordings file associated with opId:${opId} does not yet exist!`
    );
  }

  let newIdIsNotUnique = true;
  let newOpRecordingId: string;

  const opRecordings: OperationRecordings = opRecordingsFile.get('recordings');

  const opRecordingsAsArray = Object.values<
    OperationRecordingsFileVersion1['recordings'][keyof OperationRecordingsFileVersion1['recordings']]
  >(opRecordings);

  // generate a new recordingId that is unique against the existing set of recordingId's
  do {
    newOpRecordingId = shortid.generate();
    newIdIsNotUnique = opRecordingsAsArray.some(
      // eslint-disable-next-line no-loop-func
      ({ id }) => id === newOpRecordingId
    );
  } while (newIdIsNotUnique);

  const operationRecording: OperationRecording = {
    ...opRecordingWithoutId,
    id: newOpRecordingId,
  };

  const opRecordingsWithNewRecording: OperationRecordings = {
    ...opRecordings,
    [newOpRecordingId]: operationRecording,
  };

  opRecordingsFile.set('recordings', opRecordingsWithNewRecording);
  opRecordingsFile.save();

  return newOpRecordingId;
};

type CreateNewOpFile = (args: {
  opName: string;
  opRecordingWithoutId: OperationRecordingWithoutId;
  schemaId: string;
}) => Promise<{
  opId: string;
  opRecordingId: string;
}>;

export const createNewOperationRecordingsFile: CreateNewOpFile = async ({
  opName,
  opRecordingWithoutId,
  schemaId,
}) => {
  const opId = await addNewOperationToOpMappingFile({
    schemaId,
    opName,
  });

  const pathToOperationRecordingsFile = await getOperationRecordingsFilePath(
    opId
  );
  const opRecordingsFile = editJsonFile(pathToOperationRecordingsFile);

  // if this file has already been initialized, throw error
  // this is to prevent unexpected behavior
  if (doesFileExist(opRecordingsFile)) {
    throw new Error(`File for opId:${opId} already exists!`);
  }

  const opRecordingId = shortid.generate();

  const operationRecording: OperationRecording = {
    ...opRecordingWithoutId,
    id: opRecordingId,
  };

  const operationRecordingsFileContents: OperationRecordingsFileVersion1 = {
    id: opId,
    version: OPERATIONS_RECORDING_VERSION,
    recordings: {
      [opRecordingId]: operationRecording,
    },
  };

  mapObjectToJsonFile(operationRecordingsFileContents, opRecordingsFile);
  opRecordingsFile.save();

  return {
    opId,
    opRecordingId,
  };
};

type AddNewOperationToOpMappingFile = (params: {
  schemaId: string;
  opName: string;
}) => Promise<string>;

export const addNewOperationToOpMappingFile: AddNewOperationToOpMappingFile = async ({
  schemaId,
  opName,
}) => {
  const pathToOpNameToIdMappingFile = await getOpNameMappingFilePath(schemaId);
  const opsMappingFile = await editJsonFile(pathToOpNameToIdMappingFile);

  let newOpId;
  // handle the case where mappings file does not exist
  if (!doesFileExist(opsMappingFile)) {
    const initializedOpMappingFileData: OperationNameToIdMappingVersion1 = {
      version: OPERATIONS_NAME_TO_ID_MAPPING_VERSION,
      operations: {},
    };
    mapObjectToJsonFile(initializedOpMappingFileData, opsMappingFile);
    newOpId = shortid.generate();
  } else {
    let newIdIsNotUnique = true;

    // generate a new typeId that is unique against the existing set of typeId's
    do {
      newOpId = shortid.generate();
      const opMappings = Object.values<
        OperationNameToIdMappingVersion1['operations'][keyof OperationNameToIdMappingVersion1['operations']]
      >(opsMappingFile.get('operations'));
      // eslint-disable-next-line no-loop-func
      newIdIsNotUnique = opMappings.some(({ id }) => id === newOpId);
    } while (newIdIsNotUnique);
  }

  opsMappingFile.set(`operations.${opName}`, {
    name: opName,
    id: newOpId,
  });
  opsMappingFile.save();

  return newOpId;
};

type GetOperationIdFromName = (params: {
  opName: string;
  schemaId: string;
}) => Promise<string | null>;

export const getOperationIdFromName: GetOperationIdFromName = async ({
  opName,
  schemaId,
}) => {
  const opsMappingFile = await loadOperationsMappingFile(schemaId);
  if (opsMappingFile === null) {
    return null;
  }

  const opEntry = opsMappingFile.operations[opName];
  if (opEntry === undefined) {
    return null;
  }

  return opEntry.id;
};

export const getOperationFile = async (
  operationId: string
): Promise<OperationRecordingsFileVersion1 | null> => {
  if (!operationId) {
    return null;
  }
  let operationFile: OperationRecordingsFileVersion1;
  const pathToOperationFile = await getOperationRecordingsFilePath(operationId);

  try {
    operationFile = JSON.parse(
      await fs.promises.readFile(pathToOperationFile, 'utf8')
    );
  } catch (error) {
    console.error(error);
    return null;
  }

  return operationFile;
};

// allows versioning of this file structure
export type OperationNameToIdMapping = OperationNameToIdMappingVersion1;

export interface OperationNameToIdMappingVersion1 {
  version: number;
  operations: {
    [opName: string]: {
      name: string;
      id: string;
    };
  };
}

export const loadOperationsMappingFile = async (
  schemaId: string
): Promise<OperationNameToIdMappingVersion1 | null> => {
  const pathToOperationsData = await getOpNameMappingFilePath(schemaId);
  let operationsData: OperationNameToIdMappingVersion1;

  try {
    operationsData = JSON.parse(
      await fs.promises.readFile(pathToOperationsData, 'utf8')
    );
  } catch (error) {
    console.error(error);
    return null;
  }

  return operationsData;
};
