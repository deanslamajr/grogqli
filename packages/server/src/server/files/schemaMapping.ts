import editJsonFile from 'edit-json-file';
import fs from 'fs';

import {
  doesFileExist,
  mapObjectToJsonFile,
  getSchemaMappingFilePath,
  SCHEMA_MAPPINGS_FILE_VERSION,
} from './';

interface SchemaMappings {
  [schemaUrl: string]: string;
}

type SchemaMappingFile = SchemaMappingFileVersion1;
interface SchemaMappingFileVersion1 {
  version: 1;
  mappings: SchemaMappings;
}

type Add = (params: { schemaId: string; schemaUrl: string }) => Promise<void>;
export const add: Add = async ({ schemaId, schemaUrl }) => {
  const schemaMappingFilePath = await getSchemaMappingFilePath();
  const schemaMappingFile = editJsonFile(schemaMappingFilePath);

  if (doesFileExist(schemaMappingFile)) {
    const mappings = schemaMappingFile.get('mappings') as SchemaMappings;
    mappings[schemaUrl] = schemaId;
    schemaMappingFile.set('mappings', mappings);
  } else {
    const newSchemaMappingFileContents: SchemaMappingFile = {
      version: SCHEMA_MAPPINGS_FILE_VERSION,
      mappings: {
        [schemaUrl]: schemaId,
      },
    };
    mapObjectToJsonFile(newSchemaMappingFileContents, schemaMappingFile);
  }

  schemaMappingFile.save();
};

type GetIdFromUrl = (schemaUrl: string) => Promise<string | null>;
export const getIdFromUrl: GetIdFromUrl = async (schemaUrl) => {
  const schemaMappingFilePath = await getSchemaMappingFilePath();

  let schemaMappingFile: SchemaMappingFile;
  try {
    schemaMappingFile = JSON.parse(
      await fs.promises.readFile(schemaMappingFilePath, 'utf8')
    );
  } catch (error) {
    return null;
  }

  const schemaId = schemaMappingFile.mappings[schemaUrl];

  return schemaId || null;
};
