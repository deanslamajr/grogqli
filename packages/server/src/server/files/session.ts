import { Handler, HandlerState } from '@grogqli/schema';
import editJsonFile from 'edit-json-file';
import shortid from 'shortid';

import {
  doesFileExist,
  getSessionsFilePath,
  HANDLER_SESSIONS_STATE_FILE_VERSION,
  mapObjectToJsonFile,
} from './';

type HandlerSession = Omit<Handler, '__typename'>;

export type HandlerSessionStateFile = HandlerSessionStateFileVersion1;
interface HandlerSessionStateFileVersion1 {
  version: 1;
  sessions: { [sessionId: string]: HandlerSession };
}

export const create = async (name: string): Promise<Handler> => {
  const newHandler: Handler = {
    id: shortid.generate(),
    name,
    currentState: HandlerState.Recording,
  };

  const pathToSessionsFile = await getSessionsFilePath();
  const handlerSessionStateFile = editJsonFile(pathToSessionsFile);

  if (doesFileExist(handlerSessionStateFile)) {
    handlerSessionStateFile.set(`sessions.${newHandler.id}`, newHandler);
  } else {
    const newSessionStateFileContents: HandlerSessionStateFileVersion1 = {
      version: HANDLER_SESSIONS_STATE_FILE_VERSION,
      sessions: {
        [newHandler.id]: newHandler,
      },
    };
    mapObjectToJsonFile(newSessionStateFileContents, handlerSessionStateFile);
  }

  handlerSessionStateFile.save();

  return newHandler;
};
