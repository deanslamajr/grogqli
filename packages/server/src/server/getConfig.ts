import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}