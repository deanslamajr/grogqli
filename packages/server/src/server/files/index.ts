import editJsonFile from 'edit-json-file';
import fs from 'fs';
import path from 'path';
import { IntrospectionQuery } from 'graphql';

import { getConfig } from './getConfig';
import { HANDLER_SESSION_CREATED } from '../graphql/schema/resolvers/subscriptions/handlerSessionCreatedResolver';

const createDirIfDoesntExist = (dirPath: string) => {
  return fs.promises.mkdir(dirPath, { recursive: true });
};

export const doesFileExist = (file: editJsonFile.JsonEditor): boolean => {
  return Object.keys(file.read()).length > 0;
};

export const mapObjectToJsonFile = <T extends object>(
  object: T,
  file: editJsonFile.JsonEditor
): void => {
  Object.entries(object).forEach(([key, value]) => {
    file.set(key, value);
  });
};

// Directory structure
//
// /grogqli
//   /schemas
const SCHEMAS_FOLDER_NAME = 'schemas';
//     index.json
//     /<schemaId>
//       operations.json
export const OPERATIONS_NAME_TO_ID_MAPPING_FILENAME = 'operations.json';
export const OPERATIONS_NAME_TO_ID_MAPPING_VERSION = 1;
//       schema.json
const SCHEMA_FILENAME = 'schema.json';
//       types.json
const TYPES_NAME_TO_ID_MAPPING_FILENAME = 'types.json';
export const TYPES_NAME_TO_ID_MAPPING_VERSION = 1;
//   /operations
export const OPERATIONS_FOLDER_NAME = 'operations';
export const OPERATIONS_RECORDING_VERSION = 1;
//     someOperationId.json
//   /types
const TYPES_FOLDER_NAME = 'types';
//     <typeId>.json
export const TYPES_FILE_VERSION = 1;
//   /workflows
export const WORKFLOWS_FOLDER_NAME = 'workflows';
//     index.json
const WORKFLOWS_NAME_TO_ID_MAPPING_FILENAME = 'index.json';
export const WORKFLOWS_NAME_TO_ID_MAPPING_VERSION = 1;
//     <workflowId>.json
//   /local
const TEMP_FOLDER_NAME = 'local';
//     appState.json
//     /sessions
const HANDLER_SESSIONS_STATE_FOLDER_NAME = 'sessions';
//      index.json
const HANDLER_SESSION_STATE_FILENAME = 'index.json';
export const HANDLER_SESSIONS_STATE_FILE_VERSION = 1;
//      /schemas
const TEMP_SCHEMAS_FOLDER_NAME = 'schemas';
export const TEMP_SCHEMA_FILE_VERSION = 1;
//     /operations
const TEMP_OP_RECORDINGS_FOLDER_NAME = 'operations';
export const TEMP_OP_RECORDING_FILE_VERSION = 1;

// e.g. /grogqli
export const getRecordingsRootDir = async (): Promise<string> => {
  const config = await getConfig();
  const recordingsRootDir = config('recordingsSaveDirectory');
  await createDirIfDoesntExist(recordingsRootDir);
  return recordingsRootDir;
};

// e.g. /grogqli/schemas
export const getSchemasFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const schemasFolderPath = path.join(recordingsRootDir, SCHEMAS_FOLDER_NAME);
  await createDirIfDoesntExist(schemasFolderPath);
  return schemasFolderPath;
};

// e.g. /grogqli/schemas/<schemaId>/operations.json
export const getOpNameMappingFilePath = async (schemaId: string) => {
  const schemasFolderPath = await getSchemasFolderPath();

  const opNameMappingFilePath = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(opNameMappingFilePath);

  return path.join(
    opNameMappingFilePath,
    OPERATIONS_NAME_TO_ID_MAPPING_FILENAME
  );
};

// e.g. /grogqli/schemas/<schemaId>/types.json
export const getTypeNameMappingFilePath = async (schemaId: string) => {
  const schemasFolderPath = await getSchemasFolderPath();

  const typeNameMappingFilePath = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(typeNameMappingFilePath);

  return path.join(typeNameMappingFilePath, TYPES_NAME_TO_ID_MAPPING_FILENAME);
};

// e.g. /grogqli/workflows
export const getWorkflowsFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const workflowsFolderPath = path.join(
    recordingsRootDir,
    WORKFLOWS_FOLDER_NAME
  );
  await createDirIfDoesntExist(workflowsFolderPath);
  return workflowsFolderPath;
};

// e.g. /grogqli/workflows/index.json
export const getWorkflowNameMappingFilePath = async (): Promise<string> => {
  const workflowsFolderPath = await getWorkflowsFolderPath();
  return path.join(workflowsFolderPath, WORKFLOWS_NAME_TO_ID_MAPPING_FILENAME);
};

// e.g. /grogqli/workflows/<workflowId>.json
export const getWorkflowRecordingFilePath = async (
  workflowId: string
): Promise<string> => {
  const workflowsFolderPath = await getWorkflowsFolderPath();
  return path.join(workflowsFolderPath, `${workflowId}.json`);
};

// e.g. /grogqli/operations/<opId>.json
export const getOperationRecordingsFilePath = async (
  opId: string
): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const pathToOperationRecordingFile = path.join(
    recordingsRootDir,
    OPERATIONS_FOLDER_NAME
  );

  await createDirIfDoesntExist(pathToOperationRecordingFile);

  return path.join(pathToOperationRecordingFile, `${opId}.json`);
};

// e.g. /grogqli/types/<typeId>.json
export const getTypeFilePath = async (typeId: string) => {
  const recordingsRootDir = await getRecordingsRootDir();

  const typeFilePath = path.join(recordingsRootDir, TYPES_FOLDER_NAME);

  await createDirIfDoesntExist(typeFilePath);

  return path.join(typeFilePath, `${typeId}.json`);
};

// e.g. /grogqli/local/schemas
export const getTemporarySchemaRecordingsFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const schemaRecordingsPath = path.join(
    recordingsRootDir,
    TEMP_FOLDER_NAME,
    TEMP_SCHEMAS_FOLDER_NAME
  );
  await createDirIfDoesntExist(schemaRecordingsPath);
  return schemaRecordingsPath;
};

// e.g. /grogqli/local/schemas/<schemaHash>.json
type GetTemporarySchemaRecordingFilename = (
  schemaHash: string
) => Promise<string>;
export const getTemporarySchemaRecordingFilename: GetTemporarySchemaRecordingFilename = async (
  schemaHash
) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingsFolderPath();
  return path.join(tempSchemaRecordingsPath, `${schemaHash}.json`);
};

// e.g. /grogqli/local/sessions/<sessionId>
const getSessionFolderPath = async (sessionId: string) => {
  const recordingsRootDir = await getRecordingsRootDir();

  const givenSessionsPersistencePath = path.join(
    recordingsRootDir,
    TEMP_FOLDER_NAME,
    HANDLER_SESSIONS_STATE_FOLDER_NAME,
    sessionId
  );

  await createDirIfDoesntExist(givenSessionsPersistencePath);

  return givenSessionsPersistencePath;
};

// e.g. /grogqli/local/sessions/<sessionId>/index.json
export const getSessionsFilePath = async (sessionId: string) => {
  const givenSessionsPersistencePath = await getSessionFolderPath(sessionId);

  return path.join(
    givenSessionsPersistencePath,
    HANDLER_SESSION_STATE_FILENAME
  );
};

// e.g. /grogqli/local/sessions/<sessionId>/operations
export const getTempOpRecordingsFolderPath = async (
  sessionId: string
): Promise<string> => {
  const givenSessionsPersistencePath = await getSessionFolderPath(sessionId);

  const tempOpRecordingFolderPath = path.join(
    givenSessionsPersistencePath,
    TEMP_OP_RECORDINGS_FOLDER_NAME
  );

  await createDirIfDoesntExist(tempOpRecordingFolderPath);

  return tempOpRecordingFolderPath;
};

// e.g. /grogqli/local/sessions/<sessionId>/operations/<tempOpRecordingId>.json
type GetTempOpRecordingFilePath = (params: {
  sessionId: string;
  tempOpRecordingId: string;
}) => Promise<string>;
export const getTempOpRecordingFileName: GetTempOpRecordingFilePath = async ({
  sessionId,
  tempOpRecordingId,
}) => {
  const tempOpRecordingFolderPath = await getTempOpRecordingsFolderPath(
    sessionId
  );

  return path.join(tempOpRecordingFolderPath, `${tempOpRecordingId}.json`);
};

// Schema
// ***
// **
// *
export interface SchemaFile {
  id: string;
  version: number;
  introspectionQuery: IntrospectionQuery;
}

export const getSchema = async (schemaId: string): Promise<SchemaFile> => {
  const schemasFolderPath = await getSchemasFolderPath();
  const absPathToSchema = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(absPathToSchema);

  const absPathToSchemaFile = path.join(absPathToSchema, SCHEMA_FILENAME);

  return JSON.parse(await fs.promises.readFile(absPathToSchemaFile, 'utf8'));
};
