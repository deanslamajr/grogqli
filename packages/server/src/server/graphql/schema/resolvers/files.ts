import editJsonFile from 'edit-json-file';
import fs from 'fs';

import { getConfig } from '../../../getConfig';

export const getRecordingFile = async (): Promise<editJsonFile.JsonEditor> => {
  const config = await getConfig();
  const recordingsSaveDirectory = config('recordingsSaveDirectory');
  if (!fs.existsSync(recordingsSaveDirectory)) {
    fs.mkdirSync(recordingsSaveDirectory);
  }

  return editJsonFile(
    `${recordingsSaveDirectory}/${config('recordingsFilename')}.json`
  );
};
