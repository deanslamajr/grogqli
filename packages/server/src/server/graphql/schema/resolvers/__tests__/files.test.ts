import path from 'path';
import { getWorkflowById } from '../files';
import { getConfig } from '../../../../getConfig';

jest.mock('../../../../getConfig');

describe('files', () => {
  describe('getWorkflowById', () => {
    beforeEach(() => {
      const mockedGetConfig = getConfig as jest.MockedFunction<
        typeof getConfig
      >;
      mockedGetConfig.mockImplementation(async () => () => {
        return path.join(__dirname, 'grogqli');
      });
    });

    it('should return the given workflow file contents', async () => {
      const actual = await getWorkflowById('someWorkflowId');
      expect(actual).toMatchSnapshot();
    });

    describe('if workflowId is not passed', () => {
      it('should return null', async () => {
        const actual = await getWorkflowById();
        expect(actual).toBeNull();
      });
    });

    describe('if the file doesnt exist', () => {
      it('should return null', async () => {
        const actual = await getWorkflowById('workFlowDoesntExist');
        expect(actual).toBeNull();
      });
    });
  });
});
