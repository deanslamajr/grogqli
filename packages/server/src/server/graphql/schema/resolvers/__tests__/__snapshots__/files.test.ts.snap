// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files getOperationsData should return the operations file associated with the given schemaId 1`] = `
Object {
  "recordings": Object {
    "GetChores": Object {
      "opId": "llK98u0",
      "typeId": "xhg87iT1",
    },
    "someOpName": Object {
      "opId": "someOpId",
      "typeId": "someTypeId",
    },
  },
  "version": 1,
}
`;

exports[`files getTypeRecording should return the correct type recording 1`] = `
Object {
  "id": "someTypeRecordingId",
  "value": Object {
    "getAccountFromSession": "someAccountRecordingId",
    "getChores": "someGetChoresResponseRecordingId",
  },
  "version": 1,
}
`;

exports[`files getWorkflowById should return the given workflow file contents 1`] = `
Object {
  "id": "someWorkflowId",
  "name": "Some workflow",
  "recordings": Object {
    "llK98u0": Object {
      "rootTypeRecordingId": "q7f66d54",
    },
    "someOpId": Object {
      "rootTypeRecordingId": "someTypeRecordingId",
    },
  },
  "version": 1,
}
`;

exports[`files openTypeNameToIdMapping should return the operations file associated with the given schemaId 1`] = `
Object {
  "types": Object {
    "Chore": Object {
      "id": "bb333456",
      "name": "Chore",
    },
    "GetChoresResponse": Object {
      "id": "123j6tgc",
      "name": "GetChoresResponse",
    },
    "doesntHaveATypeFile": Object {
      "id": "doesntHaveATypeFileId",
      "name": "doesntHaveATypeFile",
    },
    "someTypeName": Object {
      "id": "someTypeId",
      "name": "someTypeName",
    },
  },
  "version": 1,
}
`;
